state/providers.go

```diff
@@ -0,0 +1,33 @@
+package state
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+func CheckWeb3Providers(providersURIs []string) (map[int64]string, error) {
+	if len(providersURIs) == 0 {
+		return nil, fmt.Errorf("no URIs provided")
+	}
+
+	providers := make(map[int64]string)
+	for _, uri := range providersURIs {
+		cli, err := ethclient.Dial(uri)
+		if err != nil {
+			return nil, fmt.Errorf("error dialing web3 provider uri '%s': %w", uri, err)
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		// get the chainID from the web3 endpoint
+		chainID, err := cli.ChainID(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("error getting the chainID from the web3 provider '%s': %w", uri, err)
+		}
+		providers[chainID.Int64()] = uri
```

Comment:
With this logic, if there is a web3 providers on the same chain declared twice, the last one will be the only stored.

A potential new feature that we might add in the future is to have a pool of web3 providers that automatically connect to the next if the current fail. You might take some ideas or code from the old ethservice queue we had when the Oracle was connecting to Ethereum (https://github.com/vocdoni/vocdoni-node/blob/release-azeno/service/ethevents.go#L109) and (https://github.com/vocdoni/vocdoni-node/blob/release-azeno/ethereum/handler/ethereumHandler.go#L90).

In any case, this is probably out of scope on this PR. However, we might start thinking about it; thus providers might be a `map [chainID] -> Web3ProviderHandler` 

Where Web3ProviderHandle has a list of methods, such as:

- AddProvider(string)
- NextProvider() ethClient
- GetProvider() ethClient
- String() string

The web3ProvidersHandler could make the Connect and check if the RPC endpoint actually works and is synced before returning it.

Sorry for the noise :P We might just open a new issue with this info.

Reply:
Created! https://github.com/vocdoni/census3/issues/73